### PR TITLE
Fix use of uninitialized variables in managed decimal implementation

### PR DESCRIPTION
**Pull request by jkotas**
I have noticed that decimal multiplications and divisions return bogus results intermittently while working on unrelated issues. It turns out a few places in the new managed decimal implementation use uninitialized stack variables. I have added debug-only instrumentation to catch these.

### CODE REVIEW

**Comment by jkotas** [comment]
The CoreFX tests have not hit this problem because of the stacklayout from XUnit is very predictable with  the specific stack locations all typically being zero by luck.

**Comment by jkotas** [comment]
@MeiChin-Tsai Could you please approve this change?

File: src/System.Private.CoreLib/shared/System/Decimal.DecCalc.cs [hunk 1]
```diff
@@ -156,6 +156,13 @@ private ulong Low64
                 1e80
             };
 
+            // Used to fill uninitialized stack variables with non-zero pattern in debug builds
+            [System.Diagnostics.Conditional("DEBUG")]
+            private static unsafe void Poison<T>(ref T s) where T: unmanaged
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Decimal.DecCalc.cs]: Nit: maybe DebugPoison so the debug-only-ness is clearer at call site?

**Comment by stephentoub** [comment]
> I have added debug-only instrumentation to catch these.

Would it make sense for the future to add a mode for the JIT where non-initlocals are poisoned in a similar way?

**Comment by jkotas** [comment]
Yes, that would make sense - especially if the skipinitlocals becomes more mainstream. Filled https://github.com/dotnet/coreclr/issues/25683

**Pull request merged**